### ISSUE TITLE
boundary_matrix in 7.2.2 fails

### ISSUE THREAD

**Issue by jflowernet** [prioritizr/prioritizr#257]
Having an issue with warnings from `boundary_matrix` since updating to latest CRAN release of `prioritizr` (7.2.2). Reprex, which is same as previous code @jeffreyhanson put together for #226, illustrates the specific problem which only seems to occur with the 'patches' (seamounts) planning unit structure. I have truncated the warnings, but there are lots more of the same and no boundary matrix is produced. Reverting to previous CRAN release 7.1.1 and running the same code works fine, so it does seem to be related to the update.

Any assistance much appreciated!

``` r
#boundary matrix with sf issue

library(prioritizr)
#> Loading required package: sp
#> Loading required package: raster
#> Loading required package: sf
#> Linking to GEOS 3.8.0, GDAL 3.0.4, PROJ 6.3.1; sf_use_s2() is TRUE
#> Loading required package: proto
library(terra)
#> terra 1.6.17
library(tibble)
library(dplyr)
#> 
#> Attaching package: 'dplyr'
#> The following objects are masked from 'package:terra':
#> 
#>     intersect, union
#> The following objects are masked from 'package:raster':
#> 
#>     intersect, select, union
#> The following objects are masked from 'package:stats':
#> 
#>     filter, lag
#> The following objects are masked from 'package:base':
#> 
#>     intersect, setdiff, setequal, union

# set seed for reproducibility
set.seed(500)

# define number of fish spp
n_fish <- 4

# create a simple planning grid (assuming cost of each planning unit is equal)
pu_raster <- raster(
  nrows = 200, ncols = 200,
  xmn = 0, xmx = 200, ymn = 0, ymx = 200, vals = 1
)
# create data for fish species
fish_stack <- raster::stack(lapply(seq_len(n_fish), function(x) {
  setValues(pu_raster, round(runif(ncell(pu_raster)) > 0.5))
})) %>% setNames(paste0("fish_", seq_len(n_fish)))

plot(fish_stack)                 
```

![](https://i.imgur.com/i7XCQ5P.png)

``` r

# create sea mount data
seamount_raster <-
  st_multipoint(rbind(c(20,50), c(45, 70), c(65, 90), c(100, 100), c(120, 150), c(30, 160), c(10, 20), c(15, 80), c(20, 180))) %>%
  st_buffer(7) %>%
  as("Spatial") %>%
  rasterize(pu_raster, field = 1)

# create a separate layer for each sea mount
seamount_stack <-
  seamount_raster %>%
  as("SpatRaster") %>%
  terra::patches() %>%
  as("Raster") %>%
  {lapply(sort(unique(values(.))), function(i) {
    raster::Which(. == i)
  })} %>%
  raster::stack() %>%
  setNames(paste0("seamount_", seq_len(raster::nlayers(.))))

plot(seamount_stack)
```

![](https://i.imgur.com/AdVqSrO.png)

``` r

# create grid-cell level planning unit data
## initialize data
pu_grid_data <-
  ### add in indices for planning units in raster to be organized
  tibble(id = as.list(seq_len(ncell(pu_raster)))) %>%
  ### add in cost data
  mutate(cost = raster::as.data.frame(pu_raster)[[1]]) %>%
  ### add in fish data
  bind_cols(as_tibble(raster::as.data.frame(fish_stack))) %>%
  ### add in seamount data
  mutate(seamount = 0)

pu_grid_data
#> # A tibble: 40,000 × 7
#>    id         cost fish_1 fish_2 fish_3 fish_4 seamount
#>    <list>    <dbl>  <dbl>  <dbl>  <dbl>  <dbl>    <dbl>
#>  1 <int [1]>     1      1      1      1      0        0
#>  2 <int [1]>     1      1      0      0      0        0
#>  3 <int [1]>     1      1      0      0      1        0
#>  4 <int [1]>     1      0      1      0      1        0
#>  5 <int [1]>     1      1      0      1      0        0
#>  6 <int [1]>     1      0      1      0      1        0
#>  7 <int [1]>     1      1      0      0      1        0
#>  8 <int [1]>     1      1      1      1      0        0
#>  9 <int [1]>     1      1      0      0      1        0
#> 10 <int [1]>     1      1      1      0      1        0
#> # … with 39,990 more rows

# create sea mount-level planning unit data
pu_sm_data <- lapply(names(seamount_stack), function(i) {
  ## initialize data for planning unit that corresponds to i'th sea mount
  curr_sm_pu <-
    ## add in indices for planning units in raster to be organized
    tibble(
      id = list(raster::Which(seamount_stack[[i]] > 0.5, cells = TRUE))
    ) %>%
    ## add in cost data
    mutate(
      cost = raster::cellStats(
        pu_raster * seamount_stack[[i]], "sum", na.rm = TRUE
      )[[1]]
    ) %>%
    ## calculate total amount of each non-sea mount feature in i'th sea mount pu
    bind_cols(
      raster::as.data.frame(fish_stack * seamount_stack[[i]]) %>%
        setNames(names(fish_stack)) %>%
        dplyr::summarize_all(sum)
    ) %>%
    ## add data for i'th seamount
    mutate(
      seamount = unname(
        raster::cellStats(seamount_stack[[i]], "sum", na.rm = TRUE)
      )
    )
  ## return data
  curr_sm_pu
}) %>% do.call(what = bind_rows) %>% as_tibble()

pu_sm_data
#> # A tibble: 9 × 7
#>   id           cost fish_1 fish_2 fish_3 fish_4 seamount
#>   <list>      <dbl>  <dbl>  <dbl>  <dbl>  <dbl>    <dbl>
#> 1 <int [156]>   156     77     83     93     70      156
#> 2 <int [156]>   156     78     90     76     77      156
#> 3 <int [156]>   156     81     86     80     78      156
#> 4 <int [156]>   156     78     77     71     79      156
#> 5 <int [156]>   156     80     88     77     69      156
#> 6 <int [156]>   156     73     69     74     80      156
#> 7 <int [156]>   156     82     89     87     87      156
#> 8 <int [156]>   156     66     70     78     84      156
#> 9 <int [156]>   156     67     74     84     76      156

# merge sea mount-level data and grid-cell level data togeather
pu_data <- bind_rows(pu_grid_data, pu_sm_data)

pu_data
#> # A tibble: 40,009 × 7
#>    id         cost fish_1 fish_2 fish_3 fish_4 seamount
#>    <list>    <dbl>  <dbl>  <dbl>  <dbl>  <dbl>    <dbl>
#>  1 <int [1]>     1      1      1      1      0        0
#>  2 <int [1]>     1      1      0      0      0        0
#>  3 <int [1]>     1      1      0      0      1        0
#>  4 <int [1]>     1      0      1      0      1        0
#>  5 <int [1]>     1      1      0      1      0        0
#>  6 <int [1]>     1      0      1      0      1        0
#>  7 <int [1]>     1      1      0      0      1        0
#>  8 <int [1]>     1      1      1      1      0        0
#>  9 <int [1]>     1      1      0      0      1        0
#> 10 <int [1]>     1      1      1      0      1        0
#> # … with 39,999 more rows

#get seamount planning unit ids
seamount_pu_ids <- pu_data %>% 
  slice_tail(n = nlayers(seamount_stack)) %>% 
  pull(1)

pu_grid_sf <- rasterToPolygons(pu_raster, na.rm=FALSE) %>% 
  st_as_sf() %>% 
  dplyr::select(-1)

pu_sf <-
  pu_grid_sf %>%
  bind_rows(
    lapply(seamount_pu_ids, function(i) {
      st_sf(layer = 1, geometry = st_union(pu_grid_sf[i, , drop = FALSE]))
    }) %>%
      do.call(what = bind_rows) %>%
      dplyr::select(-layer)
  )

boundary_matrix(pu_sf)
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 40009
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37413
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37213
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 40009
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37214
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37213
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 40009
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37410
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37210
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 40009
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37409
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 37209
#> Warning in vapply(y$warnings, warning, character(1)): invalid boundary data for
#> planning unit: 40009


```

<sup>Created on 2022-10-24 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by jeffreyhanson** [comment]
Thanks so much for reporting this @jflowernet! I'll look into this now and see what I can do.

**Comment by jeffreyhanson** [comment]
Sorry @jflowernet, I had a workshop 2 weeks ago and was sick most of last week - I'll try and get to this sometime this week.

**Comment by jflowernet** [comment]
Many thanks @jeffreyhanson ! And apologies because version 7.2.2 does produce a boundary matrix, I must have just presumed it was failing before. Running `identical` with the boundary matrices from `prioritizr` v 7.1.1, v 7.2.2, and the current development version returns `TRUE`. Still getting the warnings (`There were 50 or more warnings (use warnings() to see the first 50)`) under 7.2.2 and not 7.1.1 but not a major issue. Are you getting these?

I also tried the `str_tree` argument for `boundary_matrix()` as `TRUE` and `FALSE` in v7.1.1 and got no warnings either way.

**Comment by jeffreyhanson** [comment]
Sorry for the slow response - I've been swamped. My impression is that the code for generating the boundary data is working pretty much the same as before, but for some reason the warning is getting displayed. My guess is that these warnings are occuring because the planning unit data have planning units that are located spatially on top of other ones -- meaning that the total perimeter of a planning unit is not simply the sum of all the neighboring edge lengths. But I don't know yet why we don't see these warnings in 7.1.1.

**Comment by jeffreyhanson** [comment]
I'll try and get to this sometime this week to have a proper look at it.

**Comment by jflowernet** [comment]
Thanks Jeff. From my end, there's no real issue now since it is generating a boundary matrix, so I'm happy to close this unless you want to investigate the source of the warnings further?

**Comment by jeffreyhanson** [comment]
Thanks for understanding! Yeah, I'd like to keep this open until I can understand exactly what's going on.

**Comment by jeffreyhanson** [comment]
Just to give an update on this, the terra package has this new `sharedPaths()` function that can potentially be used to speed up the `boundary_matrix()` function and also make it more reliable too (meaning that the code won't make as many assumptions about the data and have to print warnings if there's a chance something might be wrong). So, when we update the package to use terra for raster data (sometime early next year), we can update the boundary matrix code too, and that will resolve this issue.

**Comment by jflowernet** [comment]
Amazing, thanks for the update! 
`terra` is great and I've started using some functions, but moving across to it totally is definitely going to be a gradual process. Good luck with the shift for `prioritizr`!

**Comment by jeffreyhanson** [comment]
This has been fixed by 12d594fe7d40d544894fe2282c62dfc0fd88c509.
